Add optional previous button

yeonkim1213 opened the pull request:
### Does this PR close any open issues?
Closes #658 

### Give a longer description of what this PR addresses and why it's needed

Implemented a Previous button under BaseIndividualComponent to give study participants a chance to go back and correct their answers. 
The previous button will always be disabled on the first page.
It will allow the user to fix answers if the next page has a previous button.

`previousButton?: boolean;`
`previousButtonText?:string;`

### Provide pictures/videos of the behavior before and after these changes (optional)

#### Previous Button

https://github.com/user-attachments/assets/99728fb9-ba9a-4882-9a7c-3fd4f651d97a

#### Previous Button - Dynamic blocks

https://github.com/user-attachments/assets/b141e745-d55c-41e6-8b82-559752095592

#### Previous Button - Allows input only if the next page has a previous button

https://github.com/user-attachments/assets/24f87cf8-e928-4925-a343-03dd7f96968a

#### Previous Button with nextButtonLocation = sidebar

https://github.com/user-attachments/assets/0cb1d636-2aa0-401b-8f46-b20c68759a33

#### Previous Button with nextButtonEnableTime = 5000

https://github.com/user-attachments/assets/72014511-49e6-4587-bdbd-ef7b5ee0da84

### Are there any additional TODOs before this PR is ready to go?
TODOs:
- [x] previousButtonLocation, previousButtonDisableTime, previousButtonEnableTime should be the same as the next button in default
- [x] Update input disabling to apply only when the next page has a previous button
- [x] Use the computed dynamic block length
- [x] Add previous button to existing demo study (Added to demo-html)
- [x] Add test
- [x] Fix disable message (Timer only enables/disables the next button)

JackWilb commented on src/components/PreviousButton.tsx:
```diff
@@ -0,0 +1,101 @@
+  // If previousButtonEnableTime is set, use that, otherwise use nextButtonEnableTime or 0 if not set
+  const previousButtonEnableTime = configInUse?.previousButtonEnableTime ? configInUse?.previousButtonEnableTime : configInUse?.nextButtonEnableTime || 0;
+  const [timer, setTimer] = useState<number | undefined>(undefined);
+  // Start a timer on first render, update timer every 100ms
+  useEffect(() => {
+    let time = 0;
+    const interval = setInterval(() => {
+      time += 100;
+      setTimer(time);
+    }, 100);
+    return () => {
+      clearInterval(interval);
+    };
+  }, []);
+  useEffect(() => {
+    if (timer && previousButtonDisableTime && timer >= previousButtonDisableTime && studyConfig.uiConfig.timeoutReject) {
+      navigate('./../__timedOut');
+    }
+  }, [previousButtonDisableTime, timer, navigate, studyConfig.uiConfig.timeoutReject]);
+
+  const buttonTimerSatisfied = useMemo(
+    () => {
+      const previousButtonDisableSatisfied = previousButtonDisableTime && timer ? timer <= previousButtonDisableTime : true;
+      const previousButtonEnableSatisfied = timer ? timer >= previousButtonEnableTime : true;
+      return previousButtonDisableSatisfied && previousButtonEnableSatisfied;
+    },
+    [previousButtonDisableTime, previousButtonEnableTime, timer],
+  );
```
We should remove this and let the next button handle it

yeonkim1213 replied:
Fixed code so that previous button uses next button's timer

JackWilb commented on src/components/response/ResponseSwitcher.tsx:
```diff
@@ -46,15 +52,25 @@ export function ResponseSwitcher({
 
   const isDisabled = useMemo(() => {
+    // Do not disable if the next page has previousButton enabled
+    if (typeof currentStep === 'number' && currentStep + 1 < flatSequence.length) {
```
This looks like it might be buggy with nested sequence from dynamic blocks

yeonkim1213 replied:
Fixed isDisabled to work in dynamic blocks

JackWilb reviewed (changes requested):
Some extra considerations and clean up

yeonkim1213 commented on src/store/hooks/useNextStep.ts:
```diff
@@ -99,7 +99,7 @@ export function useNextStep() {
     const currentWindowEvents = windowEvents && 'current' in windowEvents && windowEvents.current ? windowEvents.current.splice(0, windowEvents.current.length) : [];
 
-    if (dataCollectionEnabled && storedAnswer.endTime === -1) { // === -1 means the answer has not been saved yet
+    if (dataCollectionEnabled) {
```
I've checked replay mode but you can't update an answer since the next button is disabled - but I can update the condition to `if (dataCollectionEnabled && !isAnalysis)` to be safe

yeonkim1213 commented on src/parser/types.ts:
```diff
@@ -610,6 +610,12 @@ export interface BaseIndividualComponent {
+  previousButton?: boolean;
+  /** The text that is displayed on the previous button. */
+  previousButtonText?:string;
+  /** The location of the previous button. */
+  previousButtonLocation?: ConfigResponseBlockLocation;
```
Removed it and updated the code

yeonkim1213 commented on src/parser/types.ts:
```diff
@@ -632,6 +638,10 @@ export interface BaseIndividualComponent {
   nextButtonDisableTime?: number;
   /** A timer (in ms) after which the next button will be enabled. */
   nextButtonEnableTime?: number;
+  /** A timeout (in ms) after which the previous button will be disabled. */
+  previousButtonDisableTime?: number;
+  /** A timer (in ms) after which the previous button will be enabled. */
+  previousButtonEnableTime?: number;
```
Removed it

yeonkim1213 commented on src/store/hooks/usePreviousStep.ts:
```diff
@@ -0,0 +1,35 @@
+    const previousStep = currentStep - 1;
+
+    if (funcIndex) {
+      navigate(`/${studyId}/${encryptIndex(currentStep)}/${encryptIndex(decryptIndex(funcIndex) - 1)}${window.location.search}`);
```
Updated the condition to ` if (funcIndex && decryptIndex(funcIndex) >= 0) ` 

What I thought was
`  // Status of the previous button. If false, the previous button should be disabled
  const isPreviousDisabled = typeof currentStep !== 'number' || isAnalysis || currentStep <= 0;`
this code will disable the previous button on the first page.

JackWilb replied:
Does that make sense though? Wouldn't we want to allow going back out of the dynamic block?

yeonkim1213 replied:
Updated code to exit dynamic block when funcIndex = 0

JackWilb commented on src/components/PreviousButton.tsx:
```diff
@@ -0,0 +1,73 @@
+type Props = {
+  label?: string;
+  disabled?: boolean;
+  onClick?: null | (() => void | Promise<void>);
```
Unused

JackWilb commented on src/components/PreviousButton.tsx:
```diff
@@ -0,0 +1,73 @@
+      >
+        {label}
+      </Button>
+      {previousButtonEnableTime > 0 && timer && timer < previousButtonEnableTime && (
+        <Alert mt="md" title="Please wait" color="blue" icon={<IconInfoCircle />}>
+          The previous button will be enabled in
+          {' '}
+          {Math.ceil((previousButtonEnableTime - timer) / 1000)}
+          {' '}
+          seconds.
+        </Alert>
+      )}
+      {previousButtonDisableTime && timer && (previousButtonDisableTime - timer) < 10000 && (
+        (previousButtonDisableTime - timer) > 0
+          ? (
+            <Alert mt="md" title="Previous button disables soon" color="yellow" icon={<IconAlertTriangle />}>
+              The previous button disables in
+              {' '}
+              {Math.ceil((previousButtonDisableTime - timer) / 1000)}
+              {' '}
+              seconds.
+            </Alert>
+          ) : !studyConfig.uiConfig.timeoutReject && (
+            <Alert mt="md" title="Previous button disabled" color="red" icon={<IconAlertTriangle />}>
+              The previous button has timed out and is now disabled.
+            </Alert>
+          ))}
+    </>
```
Remove, let the next button handle it

JackWilb commented on src/components/response/ResponseBlock.tsx:
```diff
@@ -105,11 +105,10 @@ export function ResponseBlock({
 
   const showNextBtn = location === (configInUse?.nextButtonLocation || 'belowStimulus');
-
+  const showPreviousBtn = configInUse?.previousButton && showNextBtn;
```
```suggestion
```

JackWilb commented on src/components/response/ResponseBlock.tsx:
```diff
@@ -319,6 +318,7 @@ export function ResponseBlock({
             disabled={(hasCorrectAnswerFeedback && !enableNextButton) || !answerValidator.isValid()}
             label={configInUse.nextButtonText || 'Next'}
             configInUse={configInUse}
+            showPreviousBtn={showPreviousBtn}
```
```suggestion
            showPreviousBtn={configInUse.previousButton}
```

JackWilb commented on src/store/hooks/usePreviousStep.ts:
```diff
@@ -0,0 +1,56 @@
+        const nextComponent = flatSequence[currentStep + 1];
+        const nextConfig = studyConfig.components[nextComponent];
+        if (nextConfig?.previousButton) {
+          navigate(`/${studyId}/${encryptIndex(currentStep)}/${encryptIndex(0)}${window.location.search}`);
```
Use the computed dynamic block length

JackWilb replied:
This is not resolved

JackWilb commented:
Add a test

JackWilb commented:
Revert the messages. Timer now only disables the next button

yeonkim1213 commented:
Add previous button to exisitng demo study

yeonkim1213 commented:
> Add a test
Added previous button test to demo-html.spec.ts

> Revert the messages. Timer now only disables the next button
Reverted the messages so that it only applies to the next button

> Add previous button to exisitng demo study
Added to demo-html

JackWilb commented on src/components/PreviousButton.tsx:
```diff
@@ -0,0 +1,36 @@
+export function PreviousButton({
+  label = 'Previous',
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  configInUse,
```
`configInUse` is unused it seems. Let's remove this prop

yeonkim1213 replied:
Removed

yeonkim1213 replied:
Removed

JackWilb commented on src/components/NextButton.tsx:
```diff
@@ -74,45 +74,63 @@ export function NextButton({
+    return () => { };
   }, [disabled, isNextDisabled, buttonTimerSatisfied, goToNextStep, studyConfig.uiConfig.nextOnEnter]);
 
+  const buttonsDisabled = useMemo(() => disabled || isNextDisabled || !buttonTimerSatisfied, [disabled, isNextDisabled, buttonTimerSatisfied]);
```
Rename to `nextButtonDisabled` since that's what this is for
